rfd/0138-postgres-backend.md

```diff
@@ -0,0 +1,133 @@
+
+With this very simple schema, all backend operations can be expressed in a single SQL statement, making all operations take a single round-trip between the Teleport auth server and the database once the prepared statement cache is warm.
+
+As we intend to use logical decoding, we want to avoid transactions that are too big (in terms of rows changed); the only operations that affect more than one row are `DeleteRange` (which, in actual operation, is only ever used to delete a handful of rows at a time, for things like deleting a user's properties, password and MFA devices at once) and the deletion of expired items, which we'll have to do ourselves. Since logical decoding starts having performance issues when reaching the _thousands_ of rows changed at once, we only need to be careful when doing expired item deletion. Since we have no requirement that the deletion happens atomically (or in any specific order), we will take care to apply the operation across multiple transactions, limiting the amount of rows affected in each - such a mass expiration can happen, for instance, with the heartbeats of a large amount of nodes that have lost connection to the Teleport cluster.
+
+Each auth server running the backend will periodically attempt to delete expired items - at an interval tentatively set to 30 seconds - deleting small batches of expired items in different transactions until no more items can be deleted.
+
+To generate an event stream we'll connect to the database with a regular connection, create a temporary logical replication slot, then poll changes on it via the [`pg_logical_slot_get_changes`](https://www.postgresql.org/docs/current/functions-admin.html#FUNCTIONS-REPLICATION) function. As the builtin `pgoutput` logical decoding plugin cannot be used without errors via the SQL interface except in very recent versions of Postgres (11.20, 12.15, 13.11, 14.8 or 15.3, released in May 2023), we use the third-party but very popular [`wal2json`](https://github.com/eulerto/wal2json) plugin, which is available by default on the managed Postgres offerings in Azure, AWS and GCP, and is installable from the Debian/Ubuntu official packages or from the Postgres APT and YUM repos.
+
+The `pg_logical_slot_get_changes` function (with the appropriate `wal2json` options) will return one entry per change, which maps nicely to the backend event model; updates and inserts will be rendered as `OpPut` events, deletes as `OpDelete` events.
+
+### General architecture of the audit log
+
+The database schema for the audit log will be composed of just an `events` table with the appropriate indexes for searching:
+
+```pgsql
+CREATE TABLE events (
+    event_time timestamptz NOT NULL,
+    event_id uuid NOT NULL,
+    event_type text NOT NULL,
+    session_id uuid NOT NULL,
+    event_data json NOT NULL,
```

Comment:
Btw, did you consider `jsonb`? For the price of a bit larger size we get ability to effectively search inside the event data, so depends on how much we value that ability to search inside.

Reply:
> and I don't know how deep an attacker can end up controlling a rune in an audit message

We could inspect input json on error and check if it has NUL, replace and attempt rewrite 🤔

Reply:
I think that we could marshal the event to JSON and then blindly replace any occurrence of `\u0000` not preceded by a `\` with `\ufffd`; I don't think that the go json marshaler will ever emit a literal NUL and you can't ever have that sequence of characters outside of a string in any json value.

Reply:
I'm currently trying to figure out if the go json marshaler can ever output mismatched surrogate pairs, which would also be disallowed by `jsonb`, but regardless of that @fspmarshall pointed out that we might want to give ourselves the ability to change encoding away from JSON (and rely on the first character being `{` or something for old events lol) so I think I'm actually going to change the event data column to a completely opaque `bytea`.

Reply:
> @tobiaszheller, in your postgres tests for the enhanced audit log you ended up using a separate JSONB column for just some select fields you wanted to search for, was that because a GIN index on the whole event data was too unwieldy? 

Yes, it was because of index size, there are events that had 400KB size.

Reply:
I'll leave it as text-mode `json`; the guarantee that the data is well-formed JSON is still useful, and it avoids the weirdness that comes from NULs, bad unicode and numerical coercion while still allowing for server-side filtering. If we want to add certain fields to some index we should do that on separate columns anyway because of what Tobiasz mentioned.